File: frontend/src/lib/testUtils/fakers/botGameStateFaker.ts
import { faker } from "@faker-js/faker";

import { invertColor, randomizeColor } from "@/lib/utils/chessUtils";
import { createFakeMoveSnapshot } from "./moveSnapshotFaker";
import { BotGameState, BotType, GameColor } from "@/lib/apiClient";
import { createFakeMovePath } from "./movePathFaker";
import { createFakePlayer } from "./playerFaker";
import constants from "@/lib/constants";

export function createFakeBotGameState(
    overrides?: Partial<BotGameState>,
): BotGameState {
    const botColor = randomizeColor();
    const botPlayer = createFakePlayer(botColor, {
        userId: "bot:anarchybot",
        userName: "Anarchy Bot",
        countryCode: "XX",
        rating: 161660,
    });
    const player = createFakePlayer(invertColor(botColor));

    return {
        whitePlayer: botColor === GameColor.WHITE ? botPlayer : player,
        blackPlayer: botColor === GameColor.BLACK ? botPlayer : player,
        botColor,
        botType: faker.helpers.enumValue(BotType),
        sideToMove: faker.helpers.enumValue(GameColor),

        initialFen: constants.INITIAL_FEN,
        moveHistory: Array.from({
            length: faker.number.int({ min: 1, max: 5 }),
        }).map(() => createFakeMoveSnapshot()),
        legalMoves: Array.from({
            length: faker.number.int({ min: 1, max: 5 }),
        }).map(() => createFakeMovePath()),
        ...overrides,
    };
}

File: backend/Tests/AnarchyChess.Api.TestInfrastructure/Fakes/BotGameStateFaker.cs
using AnarchyChess.Api.Bots.Models;
using AnarchyChess.Api.GameSnapshot.Models;
using AnarchyChess.EngineShared;

namespace AnarchyChess.Api.TestInfrastructure.Fakes;

public class BotGameStateFaker : RecordFaker<BotGameState>
{
    public BotGameStateFaker()
    {
        StrictMode(true);
        RuleFor(x => x.WhitePlayer, f => new GamePlayerFaker(GameColor.White).Generate());
        RuleFor(x => x.BlackPlayer, f => new GamePlayerFaker(GameColor.Black).Generate());
        RuleFor(x => x.BotColor, f => f.PickRandom<GameColor>());
        RuleFor(x => x.BotType, f => f.PickRandom<BotType>());
        RuleFor(x => x.SideToMove, f => f.PickRandom<GameColor>());
        RuleFor(x => x.InitialFen, "10/10/10/10/10/10/10/10/10/10");
        RuleFor(x => x.MoveHistory, f => new MoveSnapshotFaker().Generate(f.Random.Number(1, 6)));
        RuleFor(x => x.LegalMoves, f => new MovePathFaker().Generate(f.Random.Number(1, 10)));
        RuleFor(x => x.ResultData, (GameResultData?)null);
    }
}
